### PR TITLE
Support an eventual phase for the blueprint actions

### PR DESCRIPTION
**Pull request by viveksinghggits**
## Change Overview

This PR is to support eventual phase for the blueprint actions. Even though there are problems in the main phases of an action in the blueprint the `deferPhase` would always be called. This is how we can specify that in a blueprint

The other details are discussed here #1220 

[This](https://gist.github.com/f6b1f5a18bae389bd3dcc3dddd41b65a) is an example of a blueprint is going to to look like with artifact being set from `deferPhase` and being consumed in the other action.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues

- #1221 

## Test Plan

Tests are added as part [of this PR](https://github.com/kanisterio/kanister/pull/1337).

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -478,19 +504,48 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
...
 		}
 		if err != nil {
 			reason := fmt.Sprintf("ActionSetFailed Action: %s", action.Name)
 			msg := "Failed to render output artifacts"
 			c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
+			c.executeDeferPhase(deferPhase, tp, bp, action.Name, ns, aIDX)
 			return nil
 		}
+
+		c.executeDeferPhase(deferPhase, tp, bp, action.Name, ns, aIDX)
```
Comment (pavannd1) [inline pkg/controller/controller.go:518]: If the defer phase has an output and is used in the output artifact, where/how do we add it?

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -467,6 +492,7 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 				return nil
 			}
 		} else {
+			// this updates the artifcats section of the actionset status
```
Comment (pavannd1) [inline pkg/controller/controller.go:495]:
```suggestion
			// this updates the artifacts section of the actionset status
```

**Comment by pavannd1** [comment]
Canceling build to let approved PRs go through. Will rerun once TravisCI frees up

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -418,75 +433,145 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
...
 				return nil
 			}
+
 			if err != nil {
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
 				if msg == "" {
 					msg = fmt.Sprintf("Failed to execute phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				}
 				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
+				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, err)
```
Comment (ihcsim) [inline pkg/controller/controller.go:456]: Instead of calling this in multiple places, can we use a Go `defer` like:

```diff
diff --git a/pkg/controller/controller.go b/pkg/controller/controller.go
index 21d0375c..b7660727 100644
--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -409,6 +409,10 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 	c.actionSetTombMap.Store(as.Name, t)
 	ctx = field.Context(ctx, consts.ActionsetNameKey, as.GetName())
 	t.Go(func() error {
+		var coreErr error
+		defer func() {
+			c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, coreErr)
+		}()
 		for i, p := range phases {
 			ctx = field.Context(ctx, consts.PhaseNameKey, p.Name())
 			c.logAndSuccessEvent(ctx, fmt.Sprintf("Executing phase %s", p.Name()), "Started Phase", as)
@@ -443,7 +447,7 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
 				msg := fmt.Sprintf("Failed to update phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
-				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, rErr)
+				coreErr = rErr
 				return nil
 			}
 
@@ -453,13 +457,12 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 					msg = fmt.Sprintf("Failed to execute phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				}
 				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
-				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, err)
+				coreErr = err
 				return nil
 			}
 			param.UpdatePhaseParams(ctx, tp, p.Name(), output)
 			c.logAndSuccessEvent(ctx, fmt.Sprintf("Completed phase %s", p.Name()), "Ended Phase", as)
 		}
-		c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, nil)
 		return nil
 	})
 	return nil
```

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -418,75 +433,145 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
...
+			msg := fmt.Sprintf("Failed to update ActionSet: %s", actionsetName)
+			c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
 		}
-		return nil
-	})
-	return nil
+		return
+	}
+	// Render the artifacts
+	arts, err := param.RenderArtifacts(artTpls, *tp)
```
Comment (ihcsim) [inline pkg/controller/controller.go:538]: How do phases render their artifacts? It seems like this is the only place where `param.RenderArtifacts()` is called.

File: docs/templates.rst [hunk 5]
```diff
@@ -609,3 +610,33 @@ Similarly, a phase can use Secrets as arguments:
...
+  :linenos:
+
+  type Phase struct {
+    Secrets map[string]v1.Secret
+    Output  map[string]interface{}
+  }
+
+The output artifacts generated by ``DeferPhase`` can be referenced by other actions'
+phases in the same way, the output artifact of blueprint's normal phases are
+referenced.
```
[the remark this answers is not in the document]
Comment (viveksinghggits) [reply docs/templates.rst:631]: Hi @ihcsim 
the `Phases` section doesn't specify how do we access output artifacts set by phases from other actions' phases. I am removing this sentence. Seems to be duplicate.

File: docs/tutorial.rst [hunk 6]
```diff
@@ -443,6 +443,64 @@ ConfigMap.
...
+          - pipefail
+          - -c
+          - |
+            echo "DeferPhase"
+            kando output bkpCompletedTime "10Minutes"
+  EOF
+
+
+Output from the previous phases can also be used in the ``DeferPhase`` like it
+is used in normal scenarios.
```
[the remark this answers is not in the document]
Comment (viveksinghggits) [reply docs/tutorial.rst:502]: the way we set output artifacts for `DeferPhase` is different from how we set it for normal phases. That's the reason I thought its better to have that here.
But I think I can specify just `DeferPhase` in first blueprint section as well. EDYT?

File: pkg/controller/controller.go (hunk 4 above)
Comment (viveksinghggits) [reply pkg/controller/controller.go:538]: all the phases (core and defer) set their artifact in actionset's status, and then all of them are rendered using this call.
Rendering of the artifact is covered in testcase as part of [this PR](https://github.com/kanisterio/kanister/pull/1337).

**Comment by viveksinghggits** [comment]
```
~/work/opensource/kanister/pkg (eventual-phase*) » go test                                                                                                                                                              130 ↵ vivek@vivek
{"FallbackVersion":"v0.0.0","File":"pkg/phase_test.go","Function":"anotherTestFunc","Line":200,"PreferredVersion":"v11.11.11","level":"info","msg":"Falling back to default version of the function","time":"2022-03-30T17:04:36.116072664+02:00"}
OK: 3 passed
PASS
ok  	github.com/kanisterio/kanister/pkg	0.015s

```

File: pkg/phase.go [hunk 7]
```diff
@@ -87,6 +93,47 @@ func checkSupportedArgs(supportedArgs []string, args map[string]interface{}) err
...
+	if _, ok := funcs[a.DeferPhase.Func][regVersion]; !ok {
+		if funcVersion.Equal(defaultVersion) {
+			return nil, errors.Errorf("Requested function {%s} has not been registered with version {%s}", a.DeferPhase.Func, version)
+		}
+		if _, ok := funcs[a.DeferPhase.Func][*defaultVersion]; !ok {
+			return nil, errors.Errorf("Requested function {%s} has not been registered with versions {%s} or {%s}", a.DeferPhase.Func, version, DefaultVersion)
+		}
+		log.Info().Print("Falling back to default version of the function", field.M{"Function": a.DeferPhase.Func, "PreferredVersion": version, "FallbackVersion": DefaultVersion})
+		regVersion = *defaultVersion
+	}
```
[the remark this answers is not in the document]
Comment (viveksinghggits) [reply pkg/phase.go]: I have added the unit test as well.

File: pkg/controller/controller.go (hunk 4 above)
Comment (ihcsim) [reply pkg/controller/controller.go:545]: Thanks for the explanation. Can we move line 531 to 582 into a separate function? Looks like they have nothing to do with `executeDeferPhase()`. Then we can also perform the `if deferPhase != nil` conditional earlier, say at line 417.

**Comment by viveksinghggits** [comment]
Hi @ihcsim 
can you please have another look.

File: docs/architecture.rst [hunk 8]
```diff
@@ -84,6 +85,11 @@ The definition of a ``BlueprintAction`` is:
   to the ``BlueprintAction``.
 - ``Phases`` is a required list of ``BlueprintPhases``. These phases are invoked
   in order when executing this Action.
+- ``DeferPhase`` is an optional ``BlueprintPhase`` that you want to invoke at
+  the end of the actions. ``DeferPhase`` can easily be used to do some
+  cleanup operations after the actual operation's phases/steps. Since
+  ``DeferPhase`` is used to cleanup the resources it will run even if the
+  actions' main phases are failed.
```
Comment (pavannd1) [inline docs/architecture.rst:92]:
```suggestion
- ``DeferPhase`` is an optional ``BlueprintPhase`` invoked after the
  execution of ``Phases`` defined above. A ``DeferPhase``, when defined,
  is executed regardless of the statuses of the ``Phases``.
  A ``DeferPhase`` can be used for cleanup operations at the end of an ``Action``.
```

File: docs/templates.rst [hunk 9]
```diff
@@ -609,3 +610,29 @@ Similarly, a phase can use Secrets as arguments:
...
+.. code-block:: go
+  :linenos:
+
+  type Phase struct {
+    Secrets map[string]v1.Secret
+    Output  map[string]interface{}
+  }
+
+Output artifact can be set as follows, considering the ``DeferPhase`` outputs an artifact
+with key ``key-name``.
```
Comment (pavannd1) [inline docs/templates.rst:630]:
```suggestion
Output artifact can be set as follows:
```
Since we have already used a similar example above

File: docs/tutorial.rst [hunk 10]
```diff
@@ -443,6 +443,60 @@ ConfigMap.
 If you re-execute this Kanister Action, you'll be able to see the Artifact in the
 ActionSet status.
 
+If you have the use case of ``DeferPhase``, below is how you can set the output artifact
+from the output that is being generated from ``DeferPhase``
```
Comment (pavannd1) [inline docs/tutorial.rst:447]:
```suggestion
If you use a ``DeferPhase``, you can set the output artifact
from the output generated from ``DeferPhase`` as shown below.
```

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -416,77 +444,162 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
...
+		return nil
+	})
+	return nil
+}
+
+// executeDeferPhase executes the phase that is provided as deferPhase in the blueprint actions. deferPhase,
+// if provided, must be run eventually for the blueprint action, irrespective of the other phases output.
+// Actionset status.state is going to be `complete` iff all of the phases and deferPhase is run successfully
+// otherwise respective error message would be logged and recorded as event and actionset's status.state
+// would be failed
```
Comment (pavannd1) [inline pkg/controller/controller.go:485]:
```suggestion
// executeDeferPhase executes the phase provided as a deferPhase in the blueprint action.
// deferPhase, if provided, must be run at the end of the blueprint action, irrespective of the
// statuses of the other phases. ActionSet `status.state` is going to be `complete` IFF all the
// phases and deferPhase are run successfully
// On failure, corresponding error messages are logged and recorded as events and the
// ActionSet's `status.state` is set to `failed`.
```

File: pkg/controller/controller_test.go [hunk 12]
```diff
@@ -138,7 +138,7 @@ func (s *ControllerSuite) TestWatch(c *C) {
 
 // nolint:unparam
 func (s *ControllerSuite) waitOnActionSetState(c *C, as *crv1alpha1.ActionSet, state crv1alpha1.State) error {
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```
Comment (pavannd1) [inline pkg/controller/controller_test.go:141]: Is this intentional?

File: pkg/param/param.go [hunk 13]
```diff
@@ -510,6 +511,12 @@ func UpdatePhaseParams(ctx context.Context, tp *TemplateParams, phaseName string
 	tp.Phases[phaseName].Output = output
 }
 
+// UpdateDeferPhaseParams updates the TemplateParams deferPhase output with passed output
+// this output would be generated/passed by execution of the phase
```
Comment (pavannd1) [inline pkg/param/param.go:515]:
```suggestion
// This output would be generated/passed by execution of the phase
```

**Comment by ihcsim** [comment]
So far this is working more or less as expected :+1:. One thing I notice is that it doesn't look like the validating webhook is checking the `deferPhase`. Can you confirm?

I tested with the following YAML:

```yaml
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: scale
  namespace: kanister
actions:
  scale:
    phases:
    - func: ScaleWorkload
      name: scale-down
      args:
        namespace: default
        name: nginx
        kind: deployment
        replicas: 0
    - func: ScaleWorkload
      name: error
      args:
        namespace: default
        name: non-existent
        kind: deployment
        replicas: 10
    deferPhase:
      func: ScaleWorkload
      name: scale-up
      args:
        namespace: default
        name: nginx
        kind: deployment
        replicas: 3
---
apiVersion: cr.kanister.io/v1alpha1
kind: ActionSet
metadata:
  generateName: scale-
  namespace: kanister
spec:
  actions:
  - name: scale
    blueprint: scale
    object:
      kind: Namespace
      name: default
```
When I had an invalid key in the `args` of the `deferPhase`, the validating webhook didn't seem to catch it.

**Comment by ihcsim** [comment]
I find it hard to follow the multiple places where the `Status.State` are set and `reconcile.ActionSet()` are called. Can we should refactor these code in a separate PR, with the following changes:

i. Call `reconcile.ActionSet()` just once. Right now we try to reconcile the actionset [in every iteration of the `for` loop](https://github.com/kanisterio/kanister/blob/4a5d9c5daf95c3698b9f7a39d4acc0e4952aa23b/pkg/controller/controller.go#L456-L461)
i. Within the same `for` loop, we only store the state of each phase in memory
i. Set `as.Status.State` and `as.Status.ErState.Error` just once, based on whether `coreErr` and `deferErr` are nil or not. It feels like because we are setting them in different places, we need to have some "final overrides" in `renderActionsetArtifacts()`

<details>
<summary>Here's the diff of what I think it can look like. If it makes sense, maybe we can do this in a separate PR.</summary>

```diff
diff --git a/pkg/controller/controller.go b/pkg/controller/controller.go
index aa16f6a5..51d10f10 100644
--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -383,7 +383,7 @@ func (c *Controller) handleActionSet(as *crv1alpha1.ActionSet) (err error) {
 
 // nolint:gocognit
 func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aIDX int) error {
-	action := as.Spec.Actions[aIDX]
+	action := *as.Spec.Actions[aIDX].DeepCopy()
 	c.logAndSuccessEvent(ctx, fmt.Sprintf("Executing action %s", action.Name), "Started Action", as)
 	bpName := as.Spec.Actions[aIDX].Blueprint
 	bp, err := c.crClient.CrV1alpha1().Blueprints(as.GetNamespace()).Get(ctx, bpName, v1.GetOptions{})
@@ -414,65 +414,85 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 	t.Go(func() error {
 		var coreErr error
 		defer func() {
+			// always run deferPhase after normal phases exit
 			var deferErr error
 			if deferPhase != nil {
 				deferErr = c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as)
 			}
-			c.renderActionsetArtifacts(ctx, as, aIDX, ns, name, action.Name, bp, tp, coreErr, deferErr)
-		}()
 
-		for i, p := range phases {
-			ctx = field.Context(ctx, consts.PhaseNameKey, p.Name())
-			c.logAndSuccessEvent(ctx, fmt.Sprintf("Executing phase %s", p.Name()), "Started Phase", as)
-			err = param.InitPhaseParams(ctx, c.clientset, tp, p.Name(), p.Objects())
-			var output map[string]interface{}
-			var msg string
-			if err == nil {
-				output, err = p.Exec(ctx, *bp, action.Name, *tp)
-			} else {
-				msg = fmt.Sprintf("Failed to init phase params: %#v:", as.Status.Actions[aIDX].Phases[i])
+			if coreErr == nil && deferErr == nil && len(as.Status.Actions[aIDX].Artifacts) > 0 {
+				coreErr = c.renderActionsetArtifacts(ctx, as, aIDX, ns, name, action.Name, bp, tp, coreErr, deferErr)
 			}
-			var rf func(*crv1alpha1.ActionSet) error
-			if err != nil {
-				coreErr = err
-				rf = func(ras *crv1alpha1.ActionSet) error {
+
+			// non-nil coreErr will always be added to the actionset status because
+			// it is likely to be the root cause of the actionset failure.
+			var f func(ras *crv1alpha1.ActionSet) error
+			if coreErr != nil {
+				f = func(ras *crv1alpha1.ActionSet) error {
+					ras.Status.State = crv1alpha1.StateFailed
+					ras.Status.Error = crv1alpha1.Error{
+						Message: coreErr.Error(),
+					}
+					return nil
+				}
+			} else if deferErr != nil {
+				f = func(ras *crv1alpha1.ActionSet) error {
 					ras.Status.State = crv1alpha1.StateFailed
 					ras.Status.Error = crv1alpha1.Error{
-						Message: err.Error(),
+						Message: deferErr.Error(),
 					}
-					ras.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateFailed
 					return nil
 				}
 			} else {
-				coreErr = nil
-				rf = func(ras *crv1alpha1.ActionSet) error {
-					ras.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateComplete
-					// this updates the phase output in the actionset status
-					ras.Status.Actions[aIDX].Phases[i].Output = output
+				f = func(ras *crv1alpha1.ActionSet) error {
+					ras.Status.State = crv1alpha1.StateComplete
 					return nil
 				}
 			}
 
-			if rErr := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), ns, name, rf); rErr != nil {
+			if err := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), ns, name, f); err != nil {
+				reason := fmt.Sprintf("ActionSetFailed Action: %s", action.Name)
+				msg := fmt.Sprintf("Failed to update ActionSet: %s:", name)
+				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
+				return
+			}
+		}()
+
+		for i, p := range phases {
+			ctx = field.Context(ctx, consts.PhaseNameKey, p.Name())
+			c.logAndSuccessEvent(ctx, fmt.Sprintf("Executing phase %s", p.Name()), "Started Phase", as)
+
+			if err := param.InitPhaseParams(ctx, c.clientset, tp, p.Name(), p.Objects()); err != nil {
+				as.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateFailed
+				coreErr = err
+
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
-				msg := fmt.Sprintf("Failed to update phase: %#v:", as.Status.Actions[aIDX].Phases[i])
-				c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
-				coreErr = rErr
-				return nil
+				msg := fmt.Sprintf("Failed to init phase params: %#v:", as.Status.Actions[aIDX].Phases[i])
+				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
+				break
 			}
 
+			var (
+				output map[string]interface{}
+				msg    string
+			)
+			output, err := p.Exec(ctx, *bp, action.Name, *tp)
 			if err != nil {
+				as.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateFailed
+				coreErr = err
+
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
-				if msg == "" {
-					msg = fmt.Sprintf("Failed to execute phase: %#v:", as.Status.Actions[aIDX].Phases[i])
-				}
+				msg = fmt.Sprintf("Failed to execute phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
-				coreErr = err
-				return nil
+				break
 			}
+
+			as.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateComplete
+			as.Status.Actions[aIDX].Phases[i].Output = output
 			param.UpdatePhaseParams(ctx, tp, p.Name(), output)
 			c.logAndSuccessEvent(ctx, fmt.Sprintf("Completed phase %s", p.Name()), "Ended Phase", as)
 		}
+
 		return nil
 	})
 	return nil
@@ -491,47 +511,24 @@ func (c *Controller) executeDeferPhase(ctx context.Context,
 	aIDX int,
 	as *crv1alpha1.ActionSet,
 ) error {
-	actionsetName, actionsetNS := as.GetName(), as.GetNamespace()
 	ctx = field.Context(ctx, consts.PhaseNameKey, as.Status.Actions[aIDX].DeferPhase.Name)
 	c.logAndSuccessEvent(ctx, fmt.Sprintf("Executing deferPhase %s", as.Status.Actions[aIDX].DeferPhase.Name), "Started deferPhase", as)
 
 	output, err := deferPhase.Exec(context.Background(), *bp, actionName, *tp)
-	var rf func(*crv1alpha1.ActionSet) error
 	if err != nil {
-		rf = func(as *crv1alpha1.ActionSet) error {
-			as.Status.State = crv1alpha1.StateFailed
-			as.Status.Error = crv1alpha1.Error{
-				Message: err.Error(),
-			}
-			as.Status.Actions[aIDX].DeferPhase.State = crv1alpha1.StateFailed
-			return nil
-		}
-	} else {
-		rf = func(as *crv1alpha1.ActionSet) error {
-			as.Status.Actions[aIDX].DeferPhase.State = crv1alpha1.StateComplete
-			as.Status.Actions[aIDX].DeferPhase.Output = output
-			return nil
-		}
-	}
-	var msg string
-	if rErr := reconcile.ActionSet(context.Background(), c.crClient.CrV1alpha1(), actionsetNS, actionsetName, rf); rErr != nil {
-		reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
-		msg := fmt.Sprintf("Failed to update defer phase: %#v:", as.Status.Actions[aIDX].DeferPhase)
-		c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
-		return rErr
-	}
+		as.Status.Actions[aIDX].DeferPhase.State = crv1alpha1.StateFailed
 
-	if err != nil {
 		reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
-		if msg == "" {
-			msg = fmt.Sprintf("Failed to execute defer phase: %#v:", as.Status.Actions[aIDX].DeferPhase)
-		}
+		msg := fmt.Sprintf("Failed to execute defer phase: %#v:", as.Status.Actions[aIDX].DeferPhase)
 		c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
-		return err
+	} else {
+		as.Status.Actions[aIDX].DeferPhase.State = crv1alpha1.StateComplete
+		as.Status.Actions[aIDX].DeferPhase.Output = output
+
+		c.logAndSuccessEvent(ctx, fmt.Sprintf("Completed deferPhase %s", as.Status.Actions[aIDX].DeferPhase.Name), "Ended deferPhase", as)
+		param.UpdateDeferPhaseParams(context.Background(), tp, output)
 	}
 
-	c.logAndSuccessEvent(ctx, fmt.Sprintf("Completed deferPhase %s", as.Status.Actions[aIDX].DeferPhase.Name), "Ended deferPhase", as)
-	param.UpdateDeferPhaseParams(context.Background(), tp, output)
 	return nil
 }
 
@@ -542,64 +539,16 @@ func (c *Controller) renderActionsetArtifacts(ctx context.Context,
 	bp *crv1alpha1.Blueprint,
 	tp *param.TemplateParams,
 	coreErr, deferErr error,
-) {
-	// Check if output artifacts are present
-	artTpls := as.Status.Actions[aIDX].Artifacts
-	if len(artTpls) == 0 {
-		// No artifacts, set ActionSetStatus to complete
-		if rErr := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), actionsetNS, actionsetName, func(ras *crv1alpha1.ActionSet) error {
-			if coreErr == nil && deferErr == nil {
-				ras.Status.State = crv1alpha1.StateComplete
-			} else {
-				ras.Status.State = crv1alpha1.StateFailed
-			}
-
-			return nil
-		}); rErr != nil {
-			reason := fmt.Sprintf("ActionSetFailed Action: %s", actionName)
-			msg := fmt.Sprintf("Failed to update ActionSet: %s", actionsetName)
-			c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
-		}
-		return
-	}
+) error {
 	// Render the artifacts
+	artTpls := as.Status.Actions[aIDX].Artifacts
 	arts, err := param.RenderArtifacts(artTpls, *tp)
-	var af func(*crv1alpha1.ActionSet) error
 	if err != nil {
-		af = func(ras *crv1alpha1.ActionSet) error {
-			ras.Status.State = crv1alpha1.StateFailed
-			ras.Status.Error = crv1alpha1.Error{
-				Message: err.Error(),
-			}
-			return nil
-		}
-	} else {
-		af = func(ras *crv1alpha1.ActionSet) error {
-			ras.Status.Actions[aIDX].Artifacts = arts
-			// make sure that the core phases that were run also didnt return any error
-			// and then set actionset's state to be complete
-			if coreErr == nil && deferErr == nil {
-				ras.Status.State = crv1alpha1.StateComplete
-			} else {
-				ras.Status.State = crv1alpha1.StateFailed
-			}
-			return nil
-		}
-	}
-	// Update ActionSet
-	if aErr := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), actionsetNS, actionsetName, af); aErr != nil {
-		reason := fmt.Sprintf("ActionSetFailed Action: %s", actionName)
-		msg := fmt.Sprintf("Failed to update Output Artifacts: %#v:", artTpls)
-		c.logAndErrorEvent(ctx, msg, reason, aErr, as, bp)
-		return
+		return err
 	}
 
-	if err != nil {
-		reason := fmt.Sprintf("ActionSetFailed Action: %s", actionName)
-		msg := "Failed to render output artifacts"
-		c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
-		return
-	}
+	as.Status.Actions[aIDX].Artifacts = arts
+	return nil
 }
 
 func (c *Controller) logAndErrorEvent(ctx context.Context, msg, reason string, err error, objects ...runtime.Object) {
```
</details>

**Comment by viveksinghggits** [comment]
Hi @ihcsim 
I will raise another PR to make sure that DeferPhase is also being validated by our blueprint validating logic. I have raised a separate issue to track that.

**Comment by viveksinghggits** [comment]
> `reconcile.ActionSet()` just once. R

I totally agree with you, we should refactor that to make make sure we are setting the status only once and then calling reconcile actionset.

File: pkg/controller/controller_test.go (hunk 12 above)
Comment (viveksinghggits) [reply pkg/controller/controller_test.go:141]: yes. I can see if I can make it to 20 to 15.

Comment (viveksinghggits) [reply pkg/controller/controller_test.go:141]: I will make this change as part of test PR, reverting this back to what it was earlier.

**Comment by ihcsim** [comment]
@viveksinghggits @pavannd1 I feel like we can merge this. But we should update the documentation and the upcoming release changelog to mark this feature as experimental until the following issues are resolved:

i. #1364
i. #1337
i. #1360

WDYT?

**Comment by viveksinghggits** [comment]
Hi @ihcsim 
I agree that we should merge last two PRs with this main PR. But I think we can merge refactor one later as well. 
From two issues that we want to get in with this PR, one PR is already raised, I will raise another one tomorrow.

**Comment by viveksinghggits** [comment]
Meanwhile can you please review the PR that adds the tests.

**Comment by ihcsim** [comment]
I will still like to mark this as experimental in the documentation and the release changelog until at least we add the validating webhook fix. I think we already have all the two issues discussed in this PR.

**Review by ihcsim — APPROVED** [review]
LGTM :+1:.

**Comment by viveksinghggits** [comment]
Let me raise the PR for blueprint validation and then we can merge this.

**Comment by viveksinghggits** [comment]
I have raised [a PR](https://github.com/kanisterio/kanister/pull/1370) to resolve #1360.

**Comment by viveksinghggits** [comment]
@ihcsim 
just pointing out Github actions failed but travis passed here.
